### PR TITLE
[GH-82] Add a user preference to disallow incoming Todos

### PR DESCRIPTION
**Pull request by asimsedhain**
#### Summary

Adds support for user preference to disallow incoming Todos.

Adds slash command `/todo settings block_incoming on/off` to change the preference.

When a user_1 tries to send a todo to user_2 who has this preference set to `on`, user_1 will get a message saying `@user_2 has blocked incoming todos`.

Defaults to `off`.

In the original issue, `/todo settings blockIncoming on/off` was suggested. Autocomplete would only allow for lowercase commands, thus changed it to `/todo settings block_incoming on/off`.
#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-todo/issues/82

### CODE REVIEW

File: server/command.go [hunk 1]
```diff
@@ -324,10 +341,39 @@ func (p *Plugin) runSettingsCommand(args []string, extra *model.CommandArgs) (bo
 		}
 
 		p.postCommandResponse(extra, responseMessage)
-	} else {
+
+	case "block_incoming":
+		if len(args) < 2 {
```
Comment (larkox) [inline server/command.go:346]: Maybe this can show the current setting?

File: server/command.go [hunk 2]
```diff
@@ -324,10 +341,39 @@ func (p *Plugin) runSettingsCommand(args []string, extra *model.CommandArgs) (bo
...
+
+		if err != nil {
+			responseMessage = "error saving the block_incoming preference"
+			p.API.LogDebug("runSettingsCommand: error saving the block_incoming preference", "error", err.Error())
+			return false, errors.New(responseMessage)
+		}
+
+		p.postCommandResponse(extra, responseMessage)
+	default:
 		return true, fmt.Errorf("setting `%s` not recognized", args[0])
```
Comment (larkox) [inline server/command.go:375]: Maybe this can show all current settings?

**Review by larkox — APPROVED** [review]
Awesome work! Thanks for the contribution. LGTM, just two small improvement proposals.

File: server/command.go (hunk 1 above)
Comment (asimsedhain) [reply server/command.go:346]: Thank you for the feedback! I will apply the changes.

**Comment by asimsedhain** [comment]
I have pushed the changes.

Also currently, the command `block_incoming on` results in **not** receiving todos. I think `allow_incoming` command would be more intuitive than `block_incoming`. With that, the command  `allow_incoming on`  will mean receive todos and `allow_incoming off` to mean **not** receive todos.

**Comment by larkox** [comment]
@aaronrothschild do you have any thoughts on @asimsedhain 's comment? Should we go with `allow_incoming` or `block_incoming`? Or do you prefer any other wording?

**Comment by aaronrothschild** [comment]
> @aaronrothschild do you have any thoughts on @asimsedhain 's comment? Should we go with `allow_incoming` or `block_incoming`? Or do you prefer any other wording?

Hmmm.  I think that because the user can choose to accept/decline a task, `allow incoming` alludes to it being simply automatic and you "are allowing incoming requests ***into your box***" italics is what the user may ***think***.  

Since we have auto-complete, the length of the command shouldn't be too big of a deal, so et's be more explicit:

I propose: `Receive _Task_Requests`or `Allow_Incoming_Task_Requests` as a settings on this command:
image

and the explanation text should be "Allow other Mattermost users to send a task for you to accept/decline?"
on = "Allow others to send you a Task, you can accept/decline"
off = "Block others from sending you a Task, they will see a message saying you don't accept Todo requests"

**Review by aaronrothschild — CHANGES_REQUESTED** [review]
Small changes requested in the UX of auto-complete.  

Thanks for this addition, sorry for the belated response!

**Comment by asimsedhain** [comment]
Thank you for feedback, @aaronrothschild!

**Comment by asimsedhain** [comment]
Thank you for the feedback, @aaronrothschild!

I think `allow_incoming_task_requests` is more explicit. I have attached some snippets on how it looks and pushed the requested changes. Please let me know if anything else needs to be altered.

`/todo settings`
image
`/todo settings` autocomplete
image
`/todo settings allow_incoming_task_requests` autocomplete
image
`/todo settings allow_incoming_task_requests on`
image
`/todo settings allow_incoming_task_requests off`
image
what other users see when they try to send todo to you when you have `allow_incoming_task_requests` to `off`
image

File: server/store.go [hunk 3]
```diff
@@ -415,3 +422,34 @@ func (p *Plugin) getReminderPreference(userID string) bool {
...
+// getAllowIncomingTaskRequestsPreference - gets user preference on allowing incoming task requests from other users - default value will be true if in case any error
+func (p *Plugin) getAllowIncomingTaskRequestsPreference(userID string) bool {
+	preferenceByte, appErr := p.API.KVGet(allowIncomingTaskRequestsKey(userID))
+	if appErr != nil {
+		p.API.LogError("error getting the allow incoming task requests preference, err=", appErr.Error())
+		return true
+	}
+
+	if preferenceByte == nil {
+		p.API.LogInfo(`allow incoming task requests is empty. Defaulting to "on"`)
```
Comment (hanzei) [inline server/store.go:444]: That seems a bit spammy. Can we reduce this to `LogDebug`?

File: server/store.go [hunk 4]
```diff
@@ -415,3 +422,34 @@ func (p *Plugin) getReminderPreference(userID string) bool {
...
+		return appErr
+	}
+	return nil
+}
+
+// getAllowIncomingTaskRequestsPreference - gets user preference on allowing incoming task requests from other users - default value will be true if in case any error
+func (p *Plugin) getAllowIncomingTaskRequestsPreference(userID string) bool {
+	preferenceByte, appErr := p.API.KVGet(allowIncomingTaskRequestsKey(userID))
+	if appErr != nil {
+		p.API.LogError("error getting the allow incoming task requests preference, err=", appErr.Error())
```
Comment (hanzei) [inline server/store.go:439]: Nit pick: It's idiomatic to just return the error (Maybe warped) and let the caller deal with it.

**Review by hanzei — CHANGES_REQUESTED** [review]
One nit pick and a small request. The rest of the code looks good to me :+1:

**Review by aaronrothschild — APPROVED** [review]
Looks Great to Me @asimsedhain thanks for your contribution!

**Comment by asimsedhain** [comment]
Thank you for the feedback, @hanzei! I have made the requested changes.

**Review by hanzei — APPROVED** [review]

**Comment by emilyacook** [comment]
@DHaussermann pending your review!

**Comment by DHaussermann** [comment]
@asimsedhain I'm sorry for the delays in testing this. 

I can see the autocomplete options and the UI changes that you and Aaron have discussed above. I can also see this functionality working when using the slash commands to assign ToDos to a target user. 

But I'm not seeing this work as expected when sending a ToDo via the plugin's UI. 
When I disable incoming tasks (`/todo settings allow_incoming_task_requests off `)  I still see them appear on the inbox as well as get DMs when they are received.
Screen Shot 2021-03-01 at 5 13 26 PM


Steps: 
- Login as the test user and run `/todo settings allow_incoming_task_requests off ` 
- Note ephemeral post confirms they have been disabled
- Login as a 2nd user
- Open Tod RHS and select `Add new Todo`
- Add a ToDo message and select the test user in the `Send to user` box
**Observed**: ToDo is received along with DM for the bot  
**Expected**: Possibly the same validation message seen from the slash command could appear in this modal?

**Comment by asimsedhain** [comment]
Thank you for the testing, @DHaussermann! I am sure you must be swapped with other issues.

I think I know what changes need to be made to make it work with the plugin's UI. It will take a bit of time to push the changes (about 1-2 weeks).

**Comment by lindy65** [comment]
Hi @asimsedhain - I'm helping to review stale PRs. Can you help with the status on this one please?

//cc @aaronrothschild @DHaussermann

**Comment by asimsedhain** [comment]
Sorry for the late response.

I have fixed the issue mentioned by @DHaussermann. 

Now it shows the following whenever a user sends a todo through the plugin's UI:
image

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- New functionality is working as expected
- Tested sending ToDo from the RHS
- Tested sending ToDo from a Mattermost post
- Tested sending ToDo from a slash command
- Feature now covers all the paths
- Ensured on /off setting is respected when updated 
LGTM!

Huge thanks @asimsedhain for this enhancement! Your continued effort on this is appreciated.

Added Zephyr test to cover this feature.

**Pull request merged**